File: LeetCode/Hao/273IntegerToEnglishWords.java
public class Solution {
    String[] ones = {
        "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine"
    };
    String[] teens = {
        "",  "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", 
        "Eighteen", "Nineteen"
    };
    String[] ties = {
        "", "Ten", "Twenty", "Thirty", "Forty",
        "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
    };
    
    private String getNumberLessThan1000(int m){
        String result = "";
        int h = m / 100;
        int t = m / 10 % 10;
        if (h != 0) {
            result += ones[h] + " Hundred ";
        }
        if (t > 1) {
            result += ties[t] + " ";
            if (m % 10 != 0) {
                result += ones[m % 10] + " ";
            }
        } else if (t == 1) {
            if (m % 10 != 0) {
                result += teens[m % 10] + " ";
            } else {
                result += "Ten ";
            }
        } else {
            if (m % 10 != 0) {
                result += ones[m % 10] + " ";
            }
        }
        
        return result;
    }
    private String addNumber(int num, int shift, String output) {
        int d = num / shift;
        d %= 1000;
        String result = "";
        if(d != 0) {
            result += getNumberLessThan1000(d) + output +" ";
        }
        
        return result;
    }
    
    public String numberToWords(int num) {
        if (num == 0) {
            return "Zero";
        }
        String result = "";
        
        result += addNumber(num, 1000000000, "Billion");
        result += addNumber(num, 1000000, "Million");
        result += addNumber(num, 1000, "Thousand");
        result += getNumberLessThan1000(num % 1000);
        
        return result.trim();
    }
}

File: LeetCode/Liang/273IntegerToEnglishWords.cpp
static unordered_map<int, string> dict = {
    {0, "Zero"}, {1, "One"}, {2, "Two"}, {3, "Three"}, {4, "Four"}, 
    {5, "Five"}, {6, "Six"}, {7, "Seven"}, {8, "Eight"}, {9, "Nine"},
    {10, "Ten"}, {11, "Eleven"}, {12, "Twelve"}, {13, "Thirteen"}, {14, "Fourteen"},
    {15, "Fifteen"}, {16, "Sixteen"}, {17, "Seventeen"}, {18, "Eighteen"}, {19, "Nineteen"},
    {20, "Twenty"}, {30, "Thirty"}, {40, "Forty"}, {50, "Fifty"}, {60, "Sixty"},
    {70, "Seventy"}, {80, "Eighty"}, {90, "Ninety"}
};

class Solution {
public:
    string numberToWords(int num) {
        if (num == 0) {
            return dict[num];
        }
        
        string result = "";
        result += numberUnderThousandHelper(num / 1000000000, " Billion");
        
        num %= 1000000000;
        result += numberUnderThousandHelper(num / 1000000, " Million");
        
        num %= 1000000;
        result += numberUnderThousandHelper(num / 1000, " Thousand");
        
        num %= 1000;
        result += numberUnderThousandHelper(num, "");
        
        // Eliminate front whitespace
        if (result[0] == ' ') {
            result = result.substr(1);   
        }
        return result;
    }
    
private:
    string numberUnderThousandHelper(int num, string suffix) {
        if (num == 0) return "";
        
        string result = "";
        if (num / 100) {
            result += " " + dict[num / 100];
            result += " Hundred";
        }
        
        num %= 100;
        if (num / 10 > 1) {
            result += " " + dict[(int)(num / 10) * 10];
            num %= 10;
        }
                
        if (num) {
            result += " " + dict[num];
        }
        
        return result + suffix;
    }
};
